autoload/lsp/ui/vim/completion.vim

```diff
@@ -0,0 +1,240 @@
+" vint: -ProhibitUnusedVariable
+"
+let s:context = {}
+
+function! lsp#ui#vim#completion#setup() abort
+  augroup lsp_ui_vim_completion
+    autocmd!
+    autocmd CompleteDone * call s:on_complete_done()
+  augroup END
+endfunction
+
+"
+" After CompleteDone, v:complete_item's word has been inserted into the line.
+" Yet not inserted commit characters.
+"
+" below example uses # as cursor position.
+"
+" 1. `call getbuf#`<C-x><C-o>
+" 2. select `getbufline` item.
+" 3. Insert commit characters. e.g. `(`
+" 4. fire CompleteDone, then the line is `call getbufline#`
+" 5. call feedkeys to call `s:on_complete_done_after`
+" 6. then the line is `call getbufline(#` in `s:on_complete_done_after`
+"
+function! s:on_complete_done() abort
+  let l:user_data = lsp#omni#extract_user_data_from_completed_item(v:completed_item)
+  if empty(l:user_data)
+    doautocmd User lsp_complete_done
+    return
+  endif
+
+  let s:context.line = getline('.')
+  let s:context.position = getpos('.')
+  let s:context.completed_item = copy(v:completed_item)
+  let s:context.server_name = l:user_data.server_name
+  let s:context.completion_item = l:user_data.completion_item
+  call feedkeys(printf("\<C-r>=<SNR>%d_on_complete_done_after()\<CR>", s:SID()), 'n')
+endfunction
+
+"
+" Apply textEdit or insertText(snippet) and additionalTextEdits.
+"
+function! s:on_complete_done_after() abort
+  let l:line = s:context.line
+  let l:position = s:context.position
+  let l:completed_item = s:context.completed_item
+  let l:server_name = s:context.server_name
+  let l:completion_item = s:context.completion_item
+
+  " check the commit characters are <BS> or <C-w>.
+  if strlen(getline('.')) < strlen(l:line)
+    doautocmd User lsp_complete_done
+    return ''
+  endif
+
+  " Do nothing if text_edit is disabled.
+  if !g:lsp_text_edit_enabled
+    doautocmd User lsp_complete_done
+    return ''
+  endif
+
+  let l:completion_item = s:resolve_completion_item(l:completion_item, l:server_name)
+
+  " apply textEdit or insertText(snippet).
+  let l:expand_text = s:get_expand_text(l:completed_item, l:completion_item)
+  if strlen(l:expand_text) > 0
+    call s:clear_inserted_text(
+          \   l:line,
+          \   l:position,
+          \   l:completed_item,
+          \   l:completion_item
+          \ )
+    if exists('g:lsp_snippets_expand_snippet') && len(g:lsp_snippets_expand_snippet) > 0
+      " vim-lsp-snippets expects commit characters removed.
```

Comment:
we shouldn't depend on the behavior of other plugins as much as possible. @thomasfaingnaert  Should we instead fix vim-lsp-settings?

Also might be good if you guys work combine the work for snippet support in vim since @thomasfaingnaert  has https://github.com/thomasfaingnaert/vim-lsp-snippets and @hrsh7th has https://github.com/hrsh7th/vim-vsnip.

Reply:
I'm ok for now but would be good if vim-lsp-settings gets updated too.